### PR TITLE
DDF-2341 update PdfInputTransformer for new taxonomy

### PR DESCRIPTION
**Pull request by glenhein**
#### What does this PR do?

Update PdfInputTransformer for new taxonomy. This is the second attempt, which includes the content metadata extractors.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@bdeining 
@brendan-hofmann 
@jrnorth 
@rzwiefel 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@coyotesqrl 
@jlcsmith
#### How should this be tested?
Ingest a PDF and check SOLR for non-new-taxonomy fields.
#### What are the relevant tickets?
DDF-2341
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java [hunk 1]
```diff
@@ -157,10 +256,20 @@ private Metacard transformWithExtractors(InputStream input, String id)
             } catch (InvalidPasswordException e) {
                 LOGGER.warn("Cannot transform encrypted pdf", e);
                 return initializeMetacard(id);
+
```
Comment (jrnorth) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: Extra line

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java [hunk 2]
```diff
@@ -219,101 +330,48 @@ private Metacard transformPdf(String id, PDDocument pdfDocument, String contentI
...
-            addXmlElement("modificationDate", DATE_FORMAT.format(modificationDate), metadataField);
             LOGGER.debug("PDF Modification date was: {}", DATE_FORMAT.format(modificationDate));
         }
 
-        String title = documentInformation.getTitle();
-        if (StringUtils.isNotBlank(title)) {
-            metacard.setTitle(title);
-            addXmlElement("title", title, metadataField);
+        if (usePdfTitleAsTitle) {
+            String title = documentInformation.getTitle();
```
Comment (jrnorth) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: It might be worth extracting the "get value -> check not blank -> set attribute" pattern into a private method since you use it in multiple places.

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy [hunk 3]
```diff
@@ -55,16 +74,9 @@ class PdfInputTransformerSpecTest extends Specification {
 
     void verifySamplePdfMetacard(Metacard metacard) {
         verifyThumbnailSize(metacard.thumbnail)
-        assert metacard.contentTypeName == 'pdf'
-        assert metacard.title == 'Microsoft Word - Document1'
```
Comment (jrnorth) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: Shouldn't we verify the metacard title here?

**Comment by jlcsmith** [comment]
@glenhein please rebase this PR with master

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java [hunk 4]
```diff
@@ -81,9 +85,99 @@
...
+    }
+
+    @SuppressWarnings("StatementWithEmptyBody")
+    private static byte[] generatePdfThumbnail(PDDocument pdfDocument) throws IOException {
+
+        PDFRenderer pdfRenderer = new PDFRenderer(pdfDocument);
+
+        if (pdfDocument.getNumberOfPages() < 1) {
+            /*
+             * Can there be a PDF with zero pages??? Should we throw an error or what? The
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: If the PDF has zero pages I don't think we need to worry about thumbnail generation.

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy [hunk 5]
```diff
@@ -43,9 +53,18 @@ class PdfInputTransformerSpecTest extends Specification {
     }
 
     void setup() {
+        metacardType = new MetacardTypeImpl("pdf", [
+            new AssociationsAttributes(),
+            new ContactAttributes(),
+            new MediaAttributes(),
+            new DateTimeAttributes(),
```
Comment (bdeining) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: nit : It looks like DateTime and Location aren't used on the actual `MetacardTypeImpl`, so they don't need to be added to the mock one

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java [hunk 6]
```diff
@@ -157,10 +256,20 @@ private Metacard transformWithExtractors(InputStream input, String id)
...
+
             }
         }
     }
 
+    private Metacard transformWithoutExtractors(InputStream input, String id) throws IOException {
+        try (PDDocument pdfDocument = inputStreamPDDocumentFunction.apply(input)) {
+            return transformPdf(id, pdfDocument);
+        } catch (InvalidPasswordException e) {
+            LOGGER.warn("Cannot transform encrypted pdf", e);
```
Comment (stustison) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: debug

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 4 above)
Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: Agreed. Should probably return a null or empty byte[].

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java [hunk 7]
```diff
@@ -118,18 +226,9 @@ public Metacard transform(InputStream input, String id)
...
-        } catch (InvalidPasswordException e) {
-            LOGGER.warn("Cannot transform encrypted pdf", e);
-            return initializeMetacard(id);
-        }
-    }
-
     private Metacard transformWithExtractors(InputStream input, String id)
             throws IOException, CatalogTransformerException {
-        try (FileBackedOutputStream fbos = new FileBackedOutputStream(1000000)) {
+        try (TemporaryFileBackedOutputStream fbos = new TemporaryFileBackedOutputStream()) {
```
Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: This method isn't using your new `inputStreamPDDocumentFunction`. Is that an issue?

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 6 above)
Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: There are also two `warn` methods up in `transformeWithExtractors` that should be changed to `debug`.

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 2 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: fixed

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 1 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: fixed

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy (hunk 3 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: by default, the input transformer does not set the title. There are separate test method in PdfInputTransformerTest that tests the title functionality.

**Comment by coyotesqrl** [comment]
Other than my question re: `transformWithExtractors` not using the provided func, this lgtm.

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 4 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: changed to Optional<byte[]>

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy (hunk 5 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: removed

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 6 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: changed

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 7 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: not an issue, inputStreamPDDocumentFunction is used to make the static PDDocument::load a settable function for testing

Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: It won't be an issue for your unit tests where you pass in the function?

Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: through the ctor

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy [hunk 8]
```diff
@@ -15,8 +15,17 @@ package ddf.catalog.transformer.input.pdf
...
 import ddf.catalog.content.operation.ContentMetadataExtractor
 import ddf.catalog.data.Metacard
+import ddf.catalog.data.MetacardType
 import ddf.catalog.data.impl.AttributeDescriptorImpl
 import ddf.catalog.data.impl.BasicTypes
+
+import ddf.catalog.data.impl.types.AssociationsAttributes;
+import ddf.catalog.data.impl.types.ContactAttributes;
+import ddf.catalog.data.impl.types.MediaAttributes;
+import ddf.catalog.data.impl.types.DateTimeAttributes;
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: are any of the datetime attributes being used?

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy [hunk 9]
```diff
@@ -15,8 +15,17 @@ package ddf.catalog.transformer.input.pdf
...
+import ddf.catalog.data.impl.types.LocationAttributes;
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: same question here

File: catalog/transformer/catalog-transformer-pdf/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 10]
```diff
@@ -12,9 +12,36 @@
...
 
-    <bean id="pdfTransformer" class="ddf.catalog.transformer.input.pdf.PdfInputTransformer"/>
+    <bean id="pdfMetacardType" class="ddf.catalog.data.impl.MetacardTypeImpl">
+        <argument value="pdf"/>
+        <argument>
+            <list>
+                <bean class="ddf.catalog.data.impl.types.AssociationsAttributes"/>
+                <bean class="ddf.catalog.data.impl.types.ContactAttributes"/>
+                <bean class="ddf.catalog.data.impl.types.MediaAttributes"/>
+                <bean class="ddf.catalog.data.impl.types.ValidationAttributes"/>
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-pdf/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: @glenhein - looks like the Topic category is missing here and possible DateTime.

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy (hunk 8 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: removed extra imports

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy (hunk 9 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: removed extra imports

File: catalog/transformer/catalog-transformer-pdf/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 10 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Added Topic, but I don't see DateTime being used in the transformer.

Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-pdf/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Ok.  I saw DateTime being imported somewhere so I was just double checking.

**Comment by jlcsmith** [comment]
All comments addressed.  Built and tested, LGTM 👍

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy [hunk 11]
```diff
@@ -15,8 +15,11 @@ package ddf.catalog.transformer.input.pdf
 
 import ddf.catalog.content.operation.ContentMetadataExtractor
 import ddf.catalog.data.Metacard
+import ddf.catalog.data.MetacardType
 import ddf.catalog.data.impl.AttributeDescriptorImpl
 import ddf.catalog.data.impl.BasicTypes
+import ddf.catalog.data.impl.MetacardTypeImpl
+import ddf.catalog.data.impl.types.*
```
Comment (jrnorth) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: We shouldn't be using `.*` imports.

**Comment by jrnorth** [comment]
Apart from my last comment, LGTM

File: catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy (hunk 11 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: ok, btw, that was done by the code formatter

Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: It shouldn't have if you've installed our IJ code style. It defaults to 99 imports from a package before collapsing to `*`.

Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: perhaps because it is groovy? ¯\_(ツ)_/¯

Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/test/groovy/ddf/catalog/transformer/input/pdf/PdfInputTransformerSpecTest.groovy]: AH. Yes. That would do it. We have to fix the groovy formatter. :/

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java [hunk 12]
```diff
@@ -81,9 +84,110 @@
...
+    /**
+     * @param metacardType       must be non-null
+     * @param usePdfTitleAsTitle must be non-null
+     */
+    public PdfInputTransformer(MetacardType metacardType, Boolean usePdfTitleAsTitle) {
+        this(metacardType,
+                usePdfTitleAsTitle,
+                PDDocument::load,
+                GEO_PDF_PARSER::getWktFromPDF,
+                PdfInputTransformer::generatePdfThumbnail);
```
Comment (rzwiefel) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: Just curious if you're using these method references for a particular reason besides unit testing? It definitely makes the code a little harder to follow when these are otherwise just normal methods calls. It is generally a code smell to have code changes purely for unit testing.

Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: Thinking about it a bit more, I think I agree with @rzwiefel here. Especially considering that `transformWithExtractors` isn't being tested with a unit-test defined function but getting the default, I'm not sure what's being gained.

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java [hunk 13]
```diff
@@ -219,101 +328,48 @@ private Metacard transformPdf(String id, PDDocument pdfDocument, String contentI
...
-             * original implementation assumed that a PDF would always have at least one
-             * page. That's what I've implemented here, but I don't like make those
-             * kinds of assumptions :-( But I also don't want to change the original
-             * behavior without knowing how it will impact the system.
-             */
+    private void setDateIfNotNull(Calendar calendar, MetacardImpl metacard, String attributeName,
+            String description) {
+        if (calendar != null) {
+            metacard.setAttribute(attributeName, calendar.getTime());
+            LOGGER.debug("PDF {} date was: {}", description, DATE_FORMAT.format(calendar));
```
Comment (rzwiefel) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: Don't think this debug statement is needed

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 12 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: @coyotesqrl why are you calling out transformWithExtractors? The method ref inputStreamPDDocumentFunction is called from transformWithoutExtractors as well.

Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: @rzwiefel if I don't use method refs, then I would have to create methods that can be overridden by the unit test. Either way, i need some way to support unit tests.

Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: @glenhein Ignore my comment. I see that it's been updated to use the function on line 258. I'm still not in love with the testing-specific code smell. It does make this class more complicated than it otherwise needs to be for no obvious reason without the context of the unit tests. It might be cleaner if you could extract the logic for the three replaceable operations to another class, inject a bean for that in your blueprint, and mock that class and its three methods in your unit test.

It removes the smell but does lead to some extra complexity. However, from the perspective of someone looking at your class (and this new bean) in a vacuum it won't look strange.

**Comment by jlcsmith** [comment]
@glenhein a couple comments remaining including @coyotesqrl's about the test ctor and then this should be good to go.

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 12 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: I convert the method refs to proper injected beans

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java [hunk 14]
```diff
@@ -30,59 +28,92 @@
...
+    private MetacardType metacardType;
 
-    private static final FastDateFormat DATE_FORMAT = FastDateFormat.getInstance(
-            "yyyy-MM-dd'T'HH:mm:ssZZ");
+    private boolean usePdfTitleAsTitle;
 
-    private Map<ServiceReference, ContentMetadataExtractor> contentMetadataExtractors =
-            Collections.synchronizedMap(new TreeMap<>(new ServiceComparator()));
+    /**
+     * Note: direct use of this constructor is meant for unit testing only
```
Comment (coyotesqrl) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: This doc note is no longer true, right?

**Comment by coyotesqrl** [comment]
While I personally would have skipped the interfaces and just created the beans directly that's purely a style issue. This looks _much_ better now.

Just fix the one javadoc note and this is 👍

File: catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java (hunk 14 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-pdf/src/main/java/ddf/catalog/transformer/input/pdf/PdfInputTransformer.java]: fixed

File: catalog/spatial/registry/registry-api-impl/pom.xml [hunk 15]
```diff
@@ -180,6 +180,11 @@
             <artifactId>registry-federation-admin-service</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
```
Comment (jlcsmith) [inline catalog/spatial/registry/registry-api-impl/pom.xml]: Glen - why was this added here?
